# Diabetic Hospital Readmission Prediction
## Notebook 03 — Feature Engineering

### What This Notebook Does:
- Loads the cleaned merged dataset from Notebook 01
- Calculates scientifically justified feature weights
  using Pearson correlation with the target variable
- Engineers 5 new medication features
- Engineers 4 new clinical risk features
- Engineers 3 interaction features
- Removes near-zero variance features
- Saves engineered dataset for model training

### Why Feature Engineering?
Raw columns capture isolated facts about a patient.
By combining them intelligently we create features that
better represent the underlying readmission risk patterns.

### Important Note:
NO SMOTE and NO SCALING applied in this notebook.
Both are applied in Notebook 04 on training set only
to prevent data leakage.

### Output Files:
- data/processed/diabetic_engineered.csv
  69,973 patients with engineered features

In [10]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

print("Libraries loaded!")

Libraries loaded!


In [11]:
# Load merged dataset with SDOH features
df = pd.read_csv('../data/processed/diabetic_with_sdoh.csv')

print(f'Rows:    {len(df):,}')
print(f'Columns: {df.shape[1]}')
print(f'Readmission rate: {df["readmitted_30"].mean()*100:.1f}%')
print(f'\nColumn names:')
print(df.columns.tolist())

Rows:    69,973
Columns: 75
Readmission rate: 9.0%

Column names:
['race', 'gender', 'age', 'admission_type_id', 'discharge_disposition_id', 'admission_source_id', 'time_in_hospital', 'medical_specialty', 'num_lab_procedures', 'num_procedures', 'num_medications', 'number_outpatient', 'number_emergency', 'number_inpatient', 'number_diagnoses', 'metformin', 'repaglinide', 'nateglinide', 'chlorpropamide', 'glimepiride', 'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide', 'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone', 'tolazamide', 'insulin', 'glyburide-metformin', 'glipizide-metformin', 'glimepiride-pioglitazone', 'metformin-rosiglitazone', 'metformin-pioglitazone', 'change', 'diabetesMed', 'A1c_measured', 'A1c_normal', 'A1c_elevated', 'readmitted_30', 'diag_1_Diabetes', 'diag_1_Digestive', 'diag_1_Genitourinary', 'diag_1_Injury', 'diag_1_Musculoskeletal', 'diag_1_Neoplasms', 'diag_1_Other', 'diag_1_Respiratory', 'diag_1_Unknown', 'diag_2_Diabetes', 'diag_2

We load the merged dataset saved at the end of Notebook 01. This dataset has 69,973 unique patients with 75 columns — 69 clinical and encoded features plus 6 SDOH features. We print all column names so we know exactly what features are available before we start engineering new ones.

In [12]:
# ============================================================
# Calculate Scientific Weights for med_complexity_score
# ============================================================
# Weights are calculated using Pearson correlation
# with the target variable readmitted_30
#
# Steps:
# 1. Calculate correlation of each feature with readmitted_30
# 2. Divide each by total sum = proportional weight
# 3. Features more correlated with readmission get higher weight
# ============================================================

# Select only numeric columns for correlation
numeric_df = df.select_dtypes(include='number')
correlations = numeric_df.corr()['readmitted_30'].abs()

features_for_weight = ['num_medications', 'num_procedures', 'number_diagnoses']
total_corr = sum(correlations[f] for f in features_for_weight)
weights = {f: correlations[f] / total_corr for f in features_for_weight}

print('Scientific weights based on Pearson correlation with readmitted_30:')
print(f'\n  {"Feature":<25} {"Correlation":>12} {"Weight":>10}')
print(f'  {"="*50}')
for feat, w in weights.items():
    print(f'  {feat:<25} {correlations[feat]:>12.4f} {w:>10.4f}')

print(f'\nKey insight: number_diagnoses gets highest weight ({weights["number_diagnoses"]:.4f})')
print(f'because it has the highest correlation with readmission')

Scientific weights based on Pearson correlation with readmitted_30:

  Feature                    Correlation     Weight
  num_medications                 0.0364     0.4441
  num_procedures                  0.0001     0.0018
  number_diagnoses                0.0454     0.5541

Key insight: number_diagnoses gets highest weight (0.5541)
because it has the highest correlation with readmission


We calculate feature weights using Pearson correlation with the target variable readmitted_30. Each feature's correlation is divided by the total sum of all three correlations giving proportional weights that sum to 1.0. Features with higher correlation with readmission get higher weights making the med_complexity_score scientifically justified and data-driven.

### Medication Feature Engineering

In [13]:
diabetes_meds = [
    'metformin', 'repaglinide', 'nateglinide', 'chlorpropamide',
    'glimepiride', 'acetohexamide', 'glipizide', 'glyburide',
    'tolbutamide', 'pioglitazone', 'rosiglitazone', 'acarbose',
    'miglitol', 'troglitazone', 'tolazamide', 'insulin',
    'glyburide-metformin', 'glipizide-metformin',
    'glimepiride-pioglitazone', 'metformin-rosiglitazone',
    'metformin-pioglitazone'
]

available_meds = [m for m in diabetes_meds if m in df.columns]

# 1. Total diabetes medications
df['total_diabetes_meds'] = (df[available_meds] > 0).sum(axis=1)

# 2. Medication complexity score with scientific weights
df['med_complexity_score'] = (
    df['num_medications']  * weights['num_medications'] +
    df['num_procedures']   * weights['num_procedures'] +
    df['number_diagnoses'] * weights['number_diagnoses']
)

# 3. High medication burden flag
med_threshold = df['num_medications'].quantile(0.75)
df['high_med_burden'] = (
    df['num_medications'] >= med_threshold).astype(int)

# 4. Insulin flag
df['on_insulin'] = (df['insulin'] > 0).astype(int)

# 5. Multiple drug changes
df['multiple_med_changes'] = (df['change'] > 0).astype(int)

print('Medication features engineered!')
print(f'\n  total_diabetes_meds  — mean: {df["total_diabetes_meds"].mean():.2f}')
print(f'  med_complexity_score — mean: {df["med_complexity_score"].mean():.4f}')
print(f'  high_med_burden      — {df["high_med_burden"].mean()*100:.1f}% of patients')
print(f'  on_insulin           — {df["on_insulin"].mean()*100:.1f}% of patients')
print(f'  multiple_med_changes — {df["multiple_med_changes"].mean()*100:.1f}% of patients')

Medication features engineered!

  total_diabetes_meds  — mean: 11.88
  med_complexity_score — mean: 10.9626
  high_med_burden      — 25.4% of patients
  on_insulin           — 89.5% of patients
  multiple_med_changes — 55.0% of patients


#### Medication Features — Description

##### 1. total_diabetes_meds
Count of how many diabetes medications a patient is currently taking.
Patients on more drugs have more complex, harder-to-manage conditions.
A patient on 5 diabetes drugs is clinically more complex than one on 1 drug.

##### 2. med_complexity_score
Weighted combination of 3 clinical variables using scientifically
calculated Pearson correlation weights:
  num_medications  x 0.4441
  num_procedures   x 0.0018
  number_diagnoses x 0.5541
This single score captures overall treatment complexity.
number_diagnoses gets the highest weight because it has the
strongest correlation with readmission (r=0.0454).

##### 3. high_med_burden
Binary flag — 1 if patient is in top 25% for medication count, 0 otherwise.
The 75th percentile threshold is a standard statistical cutoff for high values.
27% of patients are flagged as having high medication burden.

##### 4. on_insulin
Binary flag — 1 if patient uses insulin, 0 if not.
Insulin use signals Type 1 or severe Type 2 diabetes.
Insulin-dependent patients need closer management and have
higher readmission risk than non-insulin patients.

##### 5. multiple_med_changes
Binary flag — 1 if any diabetes medication was changed during admission.
Medication changes during admission signal an unstable condition
that needed active management — associated with higher readmission risk.

We engineer 5 new medication features. total_diabetes_meds counts how many diabetes drugs a patient is currently on — patients on more drugs have more complex conditions. med_complexity_score combines medication count, procedures, and diagnoses using the scientifically calculated weights. high_med_burden flags patients in the top 25% for medication count. on_insulin flags insulin-dependent patients who have more severe diabetes. multiple_med_changes flags patients whose medications were changed during admission indicating an unstable condition being actively managed.

### Clinical Risk Feature Engineering

In [14]:
# prior_utilization weights are based on clinical severity:

# number_inpatient  x 0.5 — highest weight
#   Inpatient stays = patient sick enough to be admitted overnight
#   Strongest signal of chronic disease severity
#   Consistent with LACE Index clinical readmission tool

# number_emergency  x 0.3 — medium weight
#   Emergency visits = sudden acute deterioration
#   More severe than routine but less than inpatient admission

# number_outpatient x 0.2 — lowest weight
#   Outpatient visits = routine scheduled appointments
#   Less urgent — could be regular checkups

# All three weights sum to 1.0
# After creating the composite we DROP the 3 components
# because keeping both composite and components is redundant
# The model would learn from the same information twice

# 1. Prior utilization score
df['prior_utilization'] = (
    df['number_inpatient']  * 0.5 +
    df['number_emergency']  * 0.3 +
    df['number_outpatient'] * 0.2
)

# Drop the 3 components to remove redundancy
df = df.drop(columns=['number_inpatient',
                       'number_emergency',
                       'number_outpatient'])

# 2. Emergency admission flag
df['emergency_admission'] = (
    df['admission_source_id'] == 7).astype(int)

# 3. Long stay flag
stay_threshold = df['time_in_hospital'].quantile(0.75)
df['long_stay'] = (
    df['time_in_hospital'] >= stay_threshold).astype(int)

# 4. High diagnosis burden
df['high_diagnosis_burden'] = (
    df['number_diagnoses'] >= 7).astype(int)

print('Clinical risk features engineered!')
print(f'\n  prior_utilization     — mean: {df["prior_utilization"].mean():.4f}')
print(f'  emergency_admission   — {df["emergency_admission"].mean()*100:.1f}% of patients')
print(f'  long_stay             — {df["long_stay"].mean()*100:.1f}% of patients')
print(f'  high_diagnosis_burden — {df["high_diagnosis_burden"].mean()*100:.1f}% of patients')
print(f'\nDropped: number_inpatient, number_emergency, number_outpatient')
print(f'Reason: prior_utilization already captures all 3 combined')

Clinical risk features engineered!

  prior_utilization     — mean: 0.1752
  emergency_admission   — 53.2% of patients
  long_stay             — 26.8% of patients
  high_diagnosis_burden — 65.2% of patients

Dropped: number_inpatient, number_emergency, number_outpatient
Reason: prior_utilization already captures all 3 combined


#### Clinical Risk Features — Description

##### 1. prior_utilization
Weighted combination of all prior healthcare usage:
  prior_utilization = number_inpatient  × 0.5
                    + number_emergency  × 0.3
                    + number_outpatient × 0.2

Weights are based on clinical severity — inpatient visits
signal the most severe prior illness, emergency visits signal
acute deterioration, outpatient visits are routine care.
Weights sum to 1.0 — consistent with the LACE Index clinical
readmission risk tool used in published medical literature.

The 3 components (number_inpatient, number_emergency,
number_outpatient) are dropped after creating this composite
to avoid redundancy in the model.

##### 2. emergency_admission
Binary flag — 1 if patient was admitted through the emergency room.
admission_source_id = 7 represents Emergency Room admission
as defined in the UCI dataset data dictionary.
Emergency admissions signal sudden acute deterioration —
higher readmission risk than planned admissions.

##### 3. long_stay
Binary flag — 1 if patient stay is in the top 25% for length.
75th percentile threshold is a standard statistical cutoff.
Longer hospital stays indicate more severe illness episodes
and complex recovery needs.

##### 4. high_diagnosis_burden
Binary flag — 1 if patient has 7 or more diagnoses.
Threshold of 7 flags patients with complex multi-disease
conditions — consistent with Charlson Comorbidity Index
which uses diagnosis count as a key readmission predictor.

We engineer 4 clinical risk features. prior_utilization combines all three prior visit types into one weighted score using clinically justified weights that sum to 1.0. Inpatient visits get the highest weight (0.5) because they signal the most severe prior illness, emergency visits get medium weight (0.3) for acute deterioration, and outpatient visits get the lowest weight (0.2) for routine care. This approach is consistent with the LACE Index — a validated clinical readmission risk tool. We then drop the three individual components because keeping both the composite and its parts would be redundant — the model would be learning from the same information twice. emergency_admission flags patients admitted through the emergency room using admission source code 7 from the UCI data dictionary. long_stay flags patients in the top 25% for hospital stay length using the 75th percentile as threshold. high_diagnosis_burden flags patients with 7 or more simultaneous diagnoses indicating complex multi-disease conditions.

### Interaction Features

In [16]:
# Interaction features capture relationships between two
# variables that neither can express alone.
# A patient with both high prior visits AND many medications
# is more complex than someone high on just one dimension.
# Multiplying two features creates a combined risk signal.

# 1. Medications x Hospital Stay
df['meds_x_stay'] = (
    df['num_medications'] * df['time_in_hospital']
)

# 2. Age x Number of Diagnoses
df['age_x_diagnoses'] = (
    df['age'] * df['number_diagnoses']
)

# 3. Obesity x Diabetes (SDOH interaction)
df['obesity_x_diabetes'] = (
    df['OBESITY'] * df['DIABETES']
)

# 4. Composite risk score
df['composite_risk'] = (
    df['high_med_burden'] +
    df['high_diagnosis_burden'] +
    df['long_stay'] +
    df['emergency_admission']
)

print('Interaction features engineered!')
print(f'\n  meds_x_stay       — mean: {df["meds_x_stay"].mean():.2f}')
print(f'  age_x_diagnoses   — mean: {df["age_x_diagnoses"].mean():.2f}')
print(f'  obesity_x_diabetes — mean: {df["obesity_x_diabetes"].mean():.2f}')
print(f'  composite_risk    — mean: {df["composite_risk"].mean():.2f}')
print(f'  composite_risk    — range: {df["composite_risk"].min()} to {df["composite_risk"].max()}')

Interaction features engineered!

  meds_x_stay       — mean: 78.32
  age_x_diagnoses   — mean: 480.80
  obesity_x_diabetes — mean: 425.99
  composite_risk    — mean: 1.71
  composite_risk    — range: 0 to 4


#### Interaction Features — Description

##### 1. meds_x_stay
num_medications × time_in_hospital

Captures patients with both many medications AND a long
hospital stay — the most complex treatment episodes.
A patient with 20 medications staying 10 days is far more
complex than someone with 5 medications staying 2 days.

##### 2. age_x_diagnoses
age × number_diagnoses

Captures older patients with many diagnoses simultaneously.
Older patients with multiple conditions are at the highest
readmission risk — age alone or diagnoses alone does not
capture this combined vulnerability.

##### 3. obesity_x_diabetes (SDOH Interaction)
OBESITY × DIABETES (community level rates)

Communities with both high obesity AND high diabetes
prevalence represent higher social risk environments.
This is our key SDOH interaction feature — combining
two social health indicators into one signal.

##### 4. composite_risk
Sum of 4 binary risk flags (range 0 to 4):
  high_med_burden + high_diagnosis_burden +
  long_stay + emergency_admission

A patient scoring 4 has ALL risk factors present.
A patient scoring 0 has none of the risk factors.
This gives a simple overall clinical risk scale.

Interaction features capture combined risk signals that individual features cannot express alone. meds_x_stay captures patients who have both many medications AND a long stay — these are the most complex treatment episodes. age_x_diagnoses captures older patients with many diagnoses — this demographic is at the highest readmission risk. obesity_x_diabetes is a SDOH interaction — communities with both high obesity and high diabetes prevalence represent higher social risk environments for diabetic patients. composite_risk adds up 4 binary risk flags into a 0 to 4 scale — a patient scoring 4 has all risk factors present simultaneously.

### Remove Near-Zero Variance Features

In [17]:
# Features where almost every patient has the same value
# add no discriminative signal to the model
# They only create noise and slow down training
# Threshold: variance < 0.01

# Check variance of all numeric columns
variance = df.select_dtypes(include='number').var()
low_var = variance[variance < 0.01].index.tolist()
cols_to_drop = [c for c in low_var if c != 'readmitted_30']

print(f'Near-zero variance features found: {len(cols_to_drop)}')
print(f'Columns to drop:')
for col in cols_to_drop:
    print(f'  {col}: variance = {variance[col]:.6f}')

# Drop near-zero variance features
df.drop(columns=cols_to_drop, inplace=True)

print(f'\nDropped {len(cols_to_drop)} near-zero variance features')
print(f'Final dataset shape: {df.shape}')

Near-zero variance features found: 16
Columns to drop:
  nateglinide: variance = 0.007654
  chlorpropamide: variance = 0.001185
  acetohexamide: variance = 0.000014
  tolbutamide: variance = 0.000243
  acarbose: variance = 0.003278
  miglitol: variance = 0.000329
  troglitazone: variance = 0.000043
  tolazamide: variance = 0.000429
  glyburide-metformin: variance = 0.007339
  glipizide-metformin: variance = 0.000100
  glimepiride-pioglitazone: variance = 0.000000
  metformin-rosiglitazone: variance = 0.000029
  metformin-pioglitazone: variance = 0.000014
  A1c_measured: variance = 0.000000
  diag_1_Unknown: variance = 0.000143
  diag_2_Unknown: variance = 0.004170

Dropped 16 near-zero variance features
Final dataset shape: (69973, 69)


After engineering all new features we check for near-zero variance columns. A feature with variance below 0.01 means almost every patient has the same value — this feature cannot help the model distinguish between readmitted and not readmitted patients because it looks identical for both groups. Keeping these columns would only add noise and increase computation time without improving predictions. We protect readmitted_30 from being dropped by excluding it from the check.


In [18]:
# Save engineered dataset
# NO SMOTE and NO SCALING applied here
# Both are applied in Notebook 04 on training set only
df.to_csv('../data/processed/diabetic_engineered.csv', index=False)

print('Engineered dataset saved!')
print(f'  Location: data/processed/diabetic_engineered.csv')
print(f'  Shape: {df.shape}')
print(f'  Rows: {len(df):,} unique patients')
print(f'  Readmission rate: {df["readmitted_30"].mean()*100:.1f}%')
print(f'\nFeature summary:')
print(f'  Original clinical features')
print(f'  27 One-Hot diagnosis columns')
print(f'  6 SDOH features')
print(f'  3 A1C binary flags')
print(f'  5 medication engineered features')
print(f'  4 clinical risk engineered features')
print(f'  4 interaction features')
print(f'\nSMOTE and scaling applied in Notebook 04 on training set only')

Engineered dataset saved!
  Location: data/processed/diabetic_engineered.csv
  Shape: (69973, 69)
  Rows: 69,973 unique patients
  Readmission rate: 9.0%

Feature summary:
  Original clinical features
  27 One-Hot diagnosis columns
  6 SDOH features
  3 A1C binary flags
  5 medication engineered features
  4 clinical risk engineered features
  4 interaction features

SMOTE and scaling applied in Notebook 04 on training set only


In [22]:
# Final Summary
print(f'\nDataset saved: data/processed/diabetic_engineered.csv')
print(f'  Rows:    {len(df):,} unique patients')
print(f'  Columns: {df.shape[1]} features')

print(f'\nNew features created:')
print(f'\nMedication Features (5):')
print(f'  total_diabetes_meds  — count of active diabetes drugs')
print(f'  med_complexity_score — weighted complexity score')
print(f'  high_med_burden      — top 25% medication flag')
print(f'  on_insulin           — insulin usage flag')
print(f'  multiple_med_changes — drug change flag')

print(f'\nClinical Risk Features (4):')
print(f'  prior_utilization     — weighted prior visits score')
print(f'  emergency_admission   — emergency room admission flag')
print(f'  long_stay             — top 25% hospital stay flag')
print(f'  high_diagnosis_burden — 7+ diagnoses flag')

print(f'\nInteraction Features (4):')
print(f'  meds_x_stay          — medications x stay length')
print(f'  age_x_diagnoses      — age x diagnosis count')
print(f'  obesity_x_diabetes   — SDOH interaction feature')
print(f'  composite_risk       — overall risk flag count (0-4)')

print(f'\nRemoved:')
print(f'  16 near-zero variance features')
print(f'  3 prior utilization components (redundant)')


Dataset saved: data/processed/diabetic_engineered.csv
  Rows:    69,973 unique patients
  Columns: 69 features

New features created:

Medication Features (5):
  total_diabetes_meds  — count of active diabetes drugs
  med_complexity_score — weighted complexity score
  high_med_burden      — top 25% medication flag
  on_insulin           — insulin usage flag
  multiple_med_changes — drug change flag

Clinical Risk Features (4):
  prior_utilization     — weighted prior visits score
  emergency_admission   — emergency room admission flag
  long_stay             — top 25% hospital stay flag
  high_diagnosis_burden — 7+ diagnoses flag

Interaction Features (4):
  meds_x_stay          — medications x stay length
  age_x_diagnoses      — age x diagnosis count
  obesity_x_diabetes   — SDOH interaction feature
  composite_risk       — overall risk flag count (0-4)

Removed:
  16 near-zero variance features
  3 prior utilization components (redundant)
